src/schema_generators/generators.cljc

```diff
@@ -198,21 +209,28 @@
    constraints is an optional mapping from schema to wrappers for the default generators,
    which can impose constraints, fix certain values, etc."
   ([schema]
-     (generator schema {}))
+   (generator schema {}))
   ([schema leaf-generators]
-     (generator schema leaf-generators {}))
+   (generator schema leaf-generators {}))
   ([schema :- Schema
     leaf-generators :- LeafGenerators
     wrappers :- GeneratorWrappers]
-     (let [leaf-generators (default-leaf-generators leaf-generators)
-           gen (fn [s params]
-                 ((or (wrappers s) identity)
-                  (or (leaf-generators s)
-                      (composite-generator (s/spec s) params))))]
-       (generators/fmap
-        (s/validator schema)
-        (gen schema {:subschema-generator gen :cache #?(:clj (java.util.IdentityHashMap.)
-                                                        :cljs (atom {}))})))))
+   (generator schema leaf-generators wrappers {}))
+  ([schema :- Schema
+    leaf-generators :- LeafGenerators
+    wrappers :- GeneratorWrappers
+    options :- SchemaOptions]
+   (let [leaf-generators (default-leaf-generators leaf-generators)
+         gen (fn [s params]
+               ((or (wrappers s) identity)
+                (or (leaf-generators s)
+                    (composite-generator (s/spec s) params))))
+         params (merge {:subschema-generator gen :cache #?(:clj (java.util.IdentityHashMap.)
+                                                           :cljs (atom {}))}
+                       (or (options schema) {:max-retries 10}))]
```

Comment:
This isn't quite right, you want to call `options` at the place where you have the schema you are trying to sample, which will typically be a subschema of `schema`.

Reply:
I've made the change on https://github.com/plumatic/schema-generators/pull/29/commits/2a48605f61ec5cab16c025346daa20c83fa047d4, is it what you have described?

Reply:
Err not quite, sorry.  

I think you want to store `options` in params as another key like `subschema-options`.

Then I'd define a helper function 

`(def max-retries [params schema] (((safe-get params :subschema-options) s) :max-retries 10))`

and wherever you have 

`(get params :max-retries 10)` 

currently you'd use `(max-retries params s)` instead.  

Does that make sense?

Reply:
I have tried what you have described. The `(get params :max-retries 10)` is part of the `CompositeGenerator` protocol implementation. When I try the following implementation:

```clj
(defn max-retries
  [schema params]
  (((macros/safe-get params :subschema-options) schema) :max-retries 10))

(defprotocol CompositeGenerator
  (composite-generator [s params]))

(extend-protocol CompositeGenerator
  schema.spec.variant.VariantSpec
  (composite-generator [s params]
    (generators/such-that
     (fn [x]
       (let [pre (.-pre ^schema.spec.variant.VariantSpec s)
             post (.-post ^schema.spec.variant.VariantSpec s)]
         (not
          (or (pre x)
              (and post (post x))))))
     (generators/one-of
       (for [o (macros/safe-get s :options)]
         (if-let [g (:guard o)]
           (generators/such-that g (sub-generator o params) (max-retries s params))
           (sub-generator o params))))
     (max-retries s params)))

  ;; TODO: this does not currently capture proper semantics of maps with
  ;; both specific keys and key schemas that can override them.
  schema.spec.collection.CollectionSpec
  (composite-generator [s params]
    (generators/such-that
     (complement (.-pre ^schema.spec.collection.CollectionSpec s))
     (generators/fmap (:konstructor s) (elements-generator (:elements s) params))
     (max-retries s params)))

  schema.spec.leaf.LeafSpec
  (composite-generator [s _]
    (macros/assert! false "You must provide a leaf generator for %s" s)))
```

This implementation does not work because the `s` parameter received in the `composite-generator` function is `(s/spec s)` which provides me no longer a `Schema`, but a `VariantSpec`. Below you can check the `options`

```clj
{#schema.core.ConditionalSchema
 {:error-symbol nil
  :preds-and-schemas [[#<clojure.core$int_QMARK_@1d8946ca>
                       #<java.lang.Class@18b4aac2 class java.lang.String>]]} {:max-retries 100}}
``` 

The `s` parameter:

```clj
#schema.spec.variant.VariantSpec
{:err-f #<schema.core.ConditionalSchema$fn__6078@3641ebdc>
 :options ({:guard #<clojure.core$int_QMARK_@1d8946ca>
            :schema #<java.lang.Class@18b4aac2 class java.lang.String>})
 :post nil
 :pre #<schema.spec.core$_PLUS_no_precondition_PLUS_@4aafe9d4>}
```

Reply:
Argh, my apologies, it's been too long and I forgot how this code worked, you had it right in https://github.com/plumatic/schema-generators/commit/2a48605f61ec5cab16c025346daa20c83fa047d4.  Sorry again!